docs/standard/io/Buffers.md

```diff
@@ -0,0 +1,401 @@
+- You must request a new buffer after calling `Advance` to continue writing more data; you cannot write to a previously acquired buffer.
+
+## [ReadOnlySequence\<T\>](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.readonlysequence-1)
+
+![image](https://user-images.githubusercontent.com/95136/64049773-cbd04600-cb2a-11e9-9d37-404488a2d6d5.png)
+
+`ReadOnlySequence<T>` is a struct that can represent a contiguous or discontiguous sequence of T. It can be constructed from:
+1. A `T[]`
+1. A `ReadOnlyMemory<T>`
+1. A pair of linked list node [`ReadOnlySequenceSegment<T>`](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.readonlysequencesegment-1?view=netstandard-2.1) and index to represent the start and end position of the sequence
+
+The 3rd representation is the most interesting one as it has performance implications on various operations on the `ReadOnlySequence<T>`:
+
+|Representation|Operation|Complexity|
+---|---|---
+|`T[]`/`ReadOnlyMemory<T>`|`Length`|`O(1)`
+|`T[]`/`ReadOnlyMemory<T>`|`GetPosition(long)`|`O(1)`
+|`T[]`/`ReadOnlyMemory<T>`|`Slice(int, int)`|`O(1)`
+|`T[]`/`ReadOnlyMemory<T>`|`Slice(SequencePostion,  SequencePostion)`|`O(1)`
+|`ReadOnlySequenceSegment<T>`|`Length`|`O(1)`
+|`ReadOnlySequenceSegment<T>`|`GetPosition(long)`|`O(number of segments)`
+|`ReadOnlySequenceSegment<T>`|`Slice(int, int)`|`O(number of segments)`
+|`ReadOnlySequenceSegment<T>`|`Slice(SequencePostion, SequencePostion)`|`O(1)`
+
+Because of this mixed representation, the `ReadOnlySequence<T>` exposes indexes as `SequencePosition` instead of an integer. A `SequencePosition` is an opaque value that represents an index into the `ReadOnlySequence<T>` where it originated. It consists of 2 parts, an integer and an object, what these 2 values represent are tied to the implementation of `ReadOnlySequence<T>`.
+
+### Accessing data
+
+The `ReadOnlySequence<T>` exposes data as an enumerable of `ReadOnlyMemory<T>`. Enumerating each of the segments can be done using a simple foreach:
+
+```C#
+long FindIndexOf(in ReadOnlySequence<byte> buffer, byte data)
+{
+    long position = 0;
+    
+    foreach (ReadOnlyMemory<byte> segment in buffer)
+    {
+        ReadOnlySpan<byte> span = segment.Span;
+        var index = span.IndexOf(data);
+        if (index != -1)
+        {
+            return position + index;
+        }
+        
+        position += span.Length;
+    }
+    
+    return -1;
+}
+```
+
+The above method searches each segment for a specific byte. If we need to keep track of each segment's `SequencePosition` then [`ReadOnlySequence.TryGet`](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.readonlysequence-1.tryget?view=netcore-3.0) would be more appropriate. Let's change the above code to return a `SequencePosition` instead of an integer. This has the added benefit of allowing the caller to avoid a second scan to get the data at a specific index.
```

Comment:
It took me a little while to work out what this mean:

> "allowing the caller to avoid a second scan"

There are two kinds of scans going on here: the `Span<byte>.IndexOf` calls, and then the scan through the segments in the sequence. My mental model was that the bulk of the scan-like work here is the `IndexOf` because that's going to touch every byte until it finds what it's looking for. (Also, so far in my work with `ReadOnlySequence<T>`, which I've been using in conjunction with pipelines, the majority have been single-segment, with multi-segment examples only appearing as you approach the end of a buffer. I've found my app gets better throughput with largish buffer sizes, ensuring that most of the time I'm in the single-segment case, so I naturally tend to think of that one as the hot path that's most perf-critical.)

So maybe qualify this as "...a second scan through the sequence's segments to get..."?